checker/src/main/java/org/checkerframework/checker/optional/OptionalVisitor.java

```diff
@@ -328,9 +360,121 @@ public void handleConditionalStatementIsPresentGet(IfTree tree) {
   public Void visitMethodInvocation(MethodInvocationTree tree, Void p) {
     handleCreationElimination(tree);
     handleNestedOptionalCreation(tree);
+    updateCalleesToCallers(tree);
     return super.visitMethodInvocation(tree, p);
   }
 
+  /**
+   * Updates {@link calleesToCallers} given a method invocation.
+   *
+   * <p>Check whether the method is in the set of methods that must be checked by the Non-Empty
+   * checker whenever a method invocation is encountered. If the method is in the set, its caller
+   * should also be checked by the Non-Empty Checker.
+   *
+   * <p>This ensures that the <i>clients</i> of any methods that must be checked by the Non-Empty
+   * Checker (i.e., methods that have preconditions related to the Non-Empty type system) are
+   * included in the set of methods to check.
+   *
+   * @param tree a method invocation tree
+   */
+  private void updateCalleesToCallers(MethodInvocationTree tree) {
+    String callee = tree.getMethodSelect().toString();
+    MethodTree caller = TreePathUtil.enclosingMethod(this.getCurrentPath());
+    if (caller != null) {
+      // Using the names of methods (as opposed to their fully-qualified name or signature) is a
+      // safe (but imprecise) over-approximation of all the methods that must be verified with the
+      // Non-Empty Checker. Overloads of methods will be detected.
+      Set<String> namesOfMethodsForNonEmptyChecker =
+          methodsToVerifyWithNonEmptyChecker.stream()
```

Comment:
`methodsToVerifyWithNonEmptyChecker` can change over time.  However, this use is safe because it is called within a method, and that method has already been added (or not).

OK, I no longer believe the above.  It is true for setting the variable within `processMethodTree()` and `updateMethodToCheckWithNonEmptyCheckerGivenPreconditions()`, but not for setting the variable within `updateMethodsToVerifyWithNonEmptyCheckerGivenNonEmptyVariable()`.  Can you add a test case that fails?

Does `processMethodTree()` need to scan the whole method looking for `@NonEmpty` annotations within its body?